### ISSUE TITLE
Clearer Documentation on using Filters from Client Side

### ISSUE THREAD

**Issue by brianafinney** [ChilliCream/hotchocolate#1730]
I know that client side is not the focus of Hot Chocolate, and perhaps this is simple frustration at failure to find proper related documentation that I need. I started with Hot Chocolate and fell in love quickly. I'm using Code First + EF, have sorting, paging, filtering, selection all set up. I have an Angular front end and I've been able to get everything set up within Angular generic enough that the service will create and structure the queries and variables proper based on the request & type. Problem is that I don't know how to structure this filter variable properly and I'm going nuts not able to find the documentation. I see that Michael Staib, shows some where clause in his .NET Oxford video around 47:52, but other than showing me a general visual, it's not helping me do this via Variables. Banana Pop gives me this information...but, not any more. I've tried moving around quotes, braces, spaces, commas, blah, blah and have yet to get any further. It would simply be GREAT to have the proper structures expected by the WHERE variables, with multiple fields, documented so I can figure out what the heck I'm supposed to be doing! I don't know if this is proper place to or not, but Stack Overflow doesn't get enough attention yet for you and Google &  my search term choices have been beyond unsuccessful. Any assistance is greatly appreciated!
 
![9](https://user-images.githubusercontent.com/28458817/79595407-a3826100-8093-11ea-878b-624fc07c9bca.PNG)

**Comment by PascalSenn** [comment]
Hi @brianafinney  
Looks like this causes quite some headache, sorry for that! 

The variables in graphql are the same as the json representation of the graphql input type.

You can use IntelliSense to autocomplete the filters in your query. 

```graphql
query ExampleQuery {
   foo(where: { bar: "asd",  baz_gt: 4}) {
      bar
      baz
   }  
}
```
You can then basically just copy and paste the values to your json variable:

```graphql
query ExampleQuery($where: FooFilter) {
   foo(where: $where) {
      bar
      baz
   }  
}
```
```json
{
    "where": { "bar": "asd",  "baz_gt": 4}
}
```


regarding AND and OR: 
All filters are by default concatenated with and. So when you have a filter like the one above, the criteria is bar AND baz_gt.  You do not have to write the AND in this case. 

for both, AND and OR you do have to provide an array value though. 
You **have to** use it like this:
```graphql
query ExampleQuery($where: FooFilter) {
   foo(where: { OR: [{bar: "asd"},{ baz_gt: 4}]}) {
      bar
      baz
   }  
}
```
```json
{
    "where": { "OR": [{"bar": "asd"},{ "baz_gt": 4}]}
}
```

You **should not** write  
```graphql
query ExampleQuery {
   foo(where: { bar: "asd",  OR: {baz_gt: 4}) {
      bar
      baz
   }  
}
```
as variables
```json
{
    "where": { "bar": "asd",  "OR": {"baz_gt": 4}}
}
```

**Comment by brianafinney** [comment]
Hi Pascal (love your name, btw)

Thank you VERY much for such a thorough and informative response. It definitely tells me I'm in the right direction, but I must have something else going on that's just not letting it happen. If I do this with one field ..I'm fine. It's only when I add more than 1. I will receive one of two errors consistently..no matter what. One error: The $where variable has invalid structure I know never even makes it to the back end before the error is thrown. The other results in an "Execution Error" regardless of how I place it in, which seems to be valid. This request DOES get to the back end...but comes back with "Unexpected Execution Error". I have an EF back end, and literally have the easiest set up. Do you have any idea what this is? Have I set up something wrong/neglected some setup?

This one: [cid:ccce8b5f-6b04-4077-94a3-177b3d2a08e1] OR this one both bring back this error:
[cid:ec857411-1d68-44ab-a1a0-9e4876d7084c]

The error stack trace looks like this:
[cid:3ba261ed-8f81-4e76-9ec3-93a2bb9db54f]


________________________________

**Comment by brianafinney** [comment]
Hi Pascal (love your name, btw)

Thank you VERY much for such a thorough and informative response. It definitely tells me I'm in the right direction, but I must have something else going on that's just not letting it happen. If I do this with one field ..I'm fine. It's only when I add more than 1. I will receive one of two errors consistently..no matter what. One error: The $where variable has invalid structure I know never even makes it to the back end before the error is thrown. The other results in an "Execution Error" regardless of how I place it in, which seems to be valid. This request DOES get to the back end...but comes back with "Unexpected Execution Error". I have an EF back end, and literally have the easiest set up. Do you have any idea what this is? Have I set up something wrong/neglected some setup?

This one:
![image](https://user-images.githubusercontent.com/28458817/79644093-8749f700-815b-11ea-92fe-5ad9c020a7ec.png)

OR this one both bring back this error:
![image (1)](https://user-images.githubusercontent.com/28458817/79644001-f83cdf00-815a-11ea-9d08-7796b0cbabc5.png)

The error stack trace looks like this:
![image (2)](https://user-images.githubusercontent.com/28458817/79644009-ff63ed00-815a-11ea-8dc7-126fc0f9a126.png)

**Comment by PascalSenn** [comment]
Thank you 🙂 

>One error: The $where variable has invalid structure I know never even makes it to the back end before the error is thrown 

In this case the validation fails, this happens when the argument is not in the correct format

> .but comes back with "Unexpected Execution Error"

Hm... there are two cases where this can happen. Either the visitor runs into an invalid state or the field you were requesting was not found on the type. Which both seem weird because validation should have already checked for this.. 
If you can provide a minimal repro of this issue, e.g. in a github repository, we can have a look at it. 
Without code and schema it is really hard to tell what the issue is

**Comment by brianafinney** [comment]
Well, I've never done of these before, but I'll do my best. I am so thankful for your assistance with this. I'll respond back when I have a repro created.

**Comment by brianafinney** [comment]
Well, holy moly batman! Here's what I can tell you...I began to create the repro and when I was looking at the packages, noticed that the versions being offered was different that the one in my project, so I uploaded my project to 11.0.0 prerelease.127 and simple hit the play button again on my query and it came back as expected. My project had 11.0.0 prerelease.121, so must have been an issue there. So, I guess it's worked out. Your first response with all the information documenting how to build these queries is HUGE for me and so very, very appreciated!!! THANK YOU PASCAL!
